File: src/lib/terminal.ts
import { isTauri } from "./storage";
import {
  NETWORK_COMMANDS,
  type RiskAssessment,
  type CommandOutput,
  type HistoryEntry,
} from "@/types/terminal";

/* ------------------------------------------------------------------ */
/* Risk classification (Critical Rule 3: network commands blocked)     */
/* ------------------------------------------------------------------ */

export function classifyRisk(command: string): RiskAssessment {
  const trimmed = command.trim();
  const firstWord = trimmed.split(/\s+/)[0] ?? "";
  const base = firstWord.split("/").pop() ?? "";

  if (NETWORK_COMMANDS.includes(base)) {
    return {
      level: "high",
      reason: `Network command "${base}" is blocked by default.`,
    };
  }

  const highPatterns = [
    /\brm\s+-rf\b/,
    /\bsudo\b/,
    /\bchmod\s+777\b/,
    /\bmkfs\b/,
    /\bdd\b/,
    /\bformat\b/,
    /\bkill\s+-9\b/,
    /\bshutdown\b/,
    /\breboot\b/,
    />\s*\/dev\/sd/,
  ];
  for (const p of highPatterns) {
    if (p.test(trimmed)) {
      return { level: "high", reason: `Dangerous pattern detected: ${p.source}` };
    }
  }

  const mediumPatterns = [
    /\brm\b/,
    /\bmv\b/,
    /\bchmod\b/,
    /\bchown\b/,
    /\bkill\b/,
    /\bpkill\b/,
    /\bsu\b/,
  ];
  for (const p of mediumPatterns) {
    if (p.test(trimmed)) {
      return { level: "medium", reason: `Moderate risk command: ${p.source}` };
    }
  }

  return { level: "low", reason: "Standard command" };
}

export function requiresConfirmation(risk: RiskAssessment): boolean {
  return risk.level === "high";
}

/* ------------------------------------------------------------------ */
/* Tauri invoke wrapper                                               */
/* ------------------------------------------------------------------ */

async function tauriInvoke<T>(cmd: string, args: Record<string, unknown>): Promise<T> {
  const { invoke } = await import("@tauri-apps/api/core");
  return invoke<T>(cmd, args);
}

/* ------------------------------------------------------------------ */
/* Public API                                                         */
/* ------------------------------------------------------------------ */

export async function executeCommand(
  command: string,
  projectRoot: string,
): Promise<CommandOutput> {
  if (isTauri()) {
    try {
      return await tauriInvoke<CommandOutput>("execute_command", {
        command,
        projectRoot,
      });
    } catch (e) {
      return {
        stdout: "",
        stderr: String(e),
        exit_code: 1,
        timed_out: false,
      };
    }
  }

  return {
    stdout: "",
    stderr: `Terminal requires a native environment (Tauri). Command: ${command}`,
    exit_code: 127,
    timed_out: false,
  };
}

export async function getHistory(): Promise<HistoryEntry[]> {
  if (isTauri()) {
    try {
      return await tauriInvoke<HistoryEntry[]>("get_history", {});
    } catch {
      return [];
    }
  }
  return [];
}

File: src-tauri/src/terminal.rs
use std::path::Path;
use std::process::Command;

use serde::{Deserialize, Serialize};
use tauri::AppHandle;

#[derive(Serialize, Deserialize)]
pub struct CommandOutput {
    pub stdout: String,
    pub stderr: String,
    pub exit_code: i32,
    pub timed_out: bool,
}

/// Commands denied by default (Critical Rule 3).
const NETWORK_COMMANDS: &[&str] = &["curl", "wget", "nc", "telnet", "ssh"];

fn classify_risk(command: &str) -> (String, String) {
    let trimmed = command.trim();
    let base = trimmed.split_whitespace().next().unwrap_or("");
    let base_name = base.rsplit('/').next().unwrap_or(base);

    if NETWORK_COMMANDS.contains(&base_name) {
        return (
            "high".to_string(),
            format!("Network command \"{base_name}\" is blocked by default."),
        );
    }

    let high_patterns = ["rm -rf", "sudo", "chmod 777", "mkfs", "dd ", "format", "shutdown", "reboot"];
    for pattern in &high_patterns {
        if trimmed.contains(pattern) {
            return (
                "high".to_string(),
                format!("Dangerous pattern detected: {pattern}"),
            );
        }
    }

    let medium_patterns = ["rm ", "mv ", "chmod", "chown", "kill", "pkill", "su "];
    for pattern in &medium_patterns {
        if trimmed.contains(pattern) {
            return (
                "medium".to_string(),
                format!("Moderate risk command: {pattern}"),
            );
        }
    }

    ("low".to_string(), "Standard command".to_string())
}

/// Appends an operation to the Security Audit Log.
fn log_operation(app: &AppHandle, operation: &str, command: &str, status: &str) {
    if let Ok(dir) = app.path().app_config_dir() {
        let _ = std::fs::create_dir_all(&dir);
        let log_path = dir.join("opencomputer-audit.log");
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_secs())
            .unwrap_or(0);
        let line = format!("{} {} {} {}\n", now, operation, status, command);
        if let Ok(mut f) = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(log_path)
        {
            let _ = f.write_all(line.as_bytes());
        }
    }
}

#[tauri::command]
pub fn execute_command(
    app: AppHandle,
    command: String,
    project_root: String,
) -> Result<CommandOutput, String> {
    let (risk_level, _reason) = classify_risk(&command);

    // Block network commands at backend level
    if risk_level == "high" {
        let base = command.trim().split_whitespace().next().unwrap_or("");
        let base_name = base.rsplit('/').next().unwrap_or(base);
        if NETWORK_COMMANDS.contains(&base_name) {
            log_operation(&app, "execute_command", &command, "blocked");
            return Err(format!(
                "Command blocked: network command \"{base_name}\" is not allowed."
            ));
        }
    }

    let cwd = Path::new(&project_root);
    if !cwd.exists() {
        return Err(format!("Project root does not exist: {project_root}"));
    }

    log_operation(&app, "execute_command", &command, "running");

    let output = Command::new("sh")
        .arg("-c")
        .arg(&command)
        .current_dir(cwd)
        .output()
        .map_err(|e| format!("Failed to execute command: {e}"))?;

    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    let stderr = String::from_utf8_lossy(&output.stderr).to_string();
    let exit_code = output.status.code().unwrap_or(-1);

    log_operation(
        &app,
        "execute_command",
        &command,
        if exit_code == 0 { "success" } else { "error" },
    );

    Ok(CommandOutput {
        stdout,
        stderr,
        exit_code,
        timed_out: false,
    })
}
